### ISSUE TITLE
Type inference isn't great with Typescript

### ISSUE THREAD

**Issue by thehappycoder** [rematch/rematch#739]
I am trying rematch for the 1st time. Unfortunately, it is not acceptable to me that the inference isn't great. Please let me know if it's a bug. I left some comments in the code.

```
import { init, RematchRootState, createModel } from "@rematch/core"

const models = {
  bookPage: createModel({
    state: {
      count: 0
    },
    reducers: {
      increment: (state, payload: number) => {
        return {
          count: state.count + payload,
          // This property isn't in the state shape and it still compiles :-(
          random: 1
        }
      }
    },
    effects: dispatch => ({
      // rootState is any :-( and setting RootState doesn't work
      incrementIn1Second: async (payload: number, rootState) => {
        await new Promise(resolve => setTimeout(resolve, 1000))
        // This compiles :-(
        dispatch.randomModel.baboom()
      }
    })
  })
}

export const store = init({
  models
})

export type Store = typeof store
export type Dispatch = typeof store.dispatch
export type RootState = RematchRootState<typeof models>

```

**Comment by 777PolarFox777** [comment]
Try to set some types explicitly:
```ts
interface BookPageState {
  count: number,
}

const models = {
  bookPage: {
    state: {
      count: 0
    } as BookPageState,
    reducers: {
      increment: (state: BookPageState, payload: number): BookPageState => {
        return {
          count: state.count + payload,
          // This is error          
          random: 1
        }
      }
    },
    effects: (dispatch: Dispatch) => ({
      incrementIn1Second: async (payload: number, rootState: RootState) => {
        await new Promise(resolve => setTimeout(resolve, 1000))
        // This will not result in error, because Rematch use Models type wich is smth like 
        // { [key: string]: ModelConfig }, so any key is allowed
        dispatch.randomModel.baboom()
      }
    })
  })
}
```

**Comment by thehappycoder** [comment]
@777PolarFox777 

> Try to set some types explicitly

That may be OK.

> This will not result in error, because Rematch use Models type wich is smth like  { [key: string]: ModelConfig }, so any key is allowed

But this is not, unfortunately.

I tried redux-toolkit's slices, which only required me to explicitly define types for actions. Works well so far.

**Comment by ShMcK** [comment]
Type inference issues may be resolved with #742
